---
title: "fig4_plot(filter dmr remove blacklist)"
author: "Zhenna Jiao"
date: "`r Sys.Date()`"
output: 
  prettydoc::html_pretty:
  theme: cayman
  highlight: github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggbiplot)
library(ggplot2)
library(corrplot)
library(magrittr)
library(RColorBrewer)
library(stringr)
library(ggpubr)
library(ggrepel)
library(readxl)
library(ComplexHeatmap)
# library(scales)
```


## set some dir
```{r}
## smp_meta_dat_df
meta_df <- readRDS("/home/jiaozhenna/methylation/project/EM_seq/data/metadata/meta_df.rds")
## fig dir
fig_proj <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/figures/fig4/"
```

## fig4C associated gene at dmr regions between GC and GP 
### do vilolinplot for all associate genes
```{r}
## import data
sample_list <- c("gp01_L14-1550","gp03_L15-31","gp04_L15-12","gp05_L15-1550",
                  "gp06_L15-13","gp07_L19-5","gp08_L19-6","gp09_L19-7","gp10_L19-8","gp11_L11-1",
                  "gp12_L11-3","gp14_L11-8","gp15_L12-1","gp17_L16-2","gp18_L16-4",
                  "gp19_L16-7","gp20_L16-8","sc01_14-Z","sc02_L14-1502","sc04_L14-24",
                  "sc05_L15-24","sc06_L15-33","sc07_L15-25","sc08_L15-1608","sc09_L19-1","sc10_L19-2","sc11_L19-3",
                  "sc12_L19-4","sc13_L11-2","sc14_L11-5","sc15_L12-2","sc16_L12-3","sc17_L12-4","sc18_L12-5",
                  "sc19_L12-6","sc20_L12-8","sc21_L16-1","sc23_L16-6")


dat_proj <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/filter_sample/dmr_file_delta0.15/"
## change data format
sample_DMR_meth <- read.table(paste0(dat_proj,"dmrs_region_list.bed"),col.names = "chr_region")
sample_DMR_cov <- read.table(paste0(dat_proj,"dmrs_region_list.bed"),col.names = "chr_region")

for (i_sample in sample_list){
  a <- read.table(paste0(dat_proj,i_sample,"_dmr_meth.bed"),col.name = c("chromosome","start","end","mod","unmod","meth_level"))

 sample_DMR_meth[[i_sample]] <- a$meth_level
 sample_DMR_cov[[i_sample]] <- a$mod+a$unmod
}

## import dmr_region genes
dmr_ann <- read.csv("/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/meth_merge_change.csv")
## change data format
dmr_ann[, c(1:5)] <- lapply(dmr_ann[,c(1:5)], as.character)
dmr_ann$gene_region <-  apply(dmr_ann[, c(1:5)], 1, function(x) paste(x, collapse = "_"))

dmr_ann$chr_region <- apply(dmr_ann[, c(1:3)], 1, function(x) paste(x, collapse = "_"))

dmr_ann_df <- dmr_ann[,c(45,44,6:43)]
## merge smp_meth_data into gene annotation
merge_smp_meth_ann <- merge(dmr_ann_df,sample_DMR_meth,by="chr_region")
smp_meth_ann <- merge_smp_meth_ann[,c(2,41:78)]

##merge smp_cov_data into gene annotation
merge_smp_cov_ann <- merge(dmr_ann_df,sample_DMR_cov,by="chr_region")
smp_cov_ann <- merge_smp_cov_ann[,c(2,41:78)]

## sample_meth
data <- smp_meth_ann
rownames(data) <- as.character(data$gene_region)
data <- data[, -1]
tr_data <- t(data)
rownames(tr_data) <- colnames(data)
colnames(tr_data) <- rownames(data)
tr_data_meth <- as.data.frame(tr_data)
# tr_data_meth$smp_type <- meta_df$sc_type
## sample_cov
data_cov <- smp_cov_ann
rownames(data_cov) <- as.character(data_cov$gene_region)
data_cov <- data_cov[, -1]
tr_data_co <- t(data_cov)
rownames(tr_data_co) <- colnames(data_cov)
colnames(tr_data_co) <- rownames(data_cov)
tr_data_cov<- as.data.frame(tr_data_co)
# tr_data_cov$smp_type <- meta_df$sc_type

## write a loop to do plot 

common_cols <- intersect(names(tr_data_cov), names(tr_data_meth))

count=0
p_adjust <- list()
pvalue_list <- list()
## 
for (col in common_cols) {
  df <- data.frame(x = tr_data_cov[[col]], y = tr_data_meth[[col]]) 
  # df$status <- str_sub(rownames(tr_data_cov),1,2)
  # df$smp <- str_sub(rownames(tr_data_cov),1,4)
  df$status <- meta_df$sc_type%>%as.factor()
  df$smp <- str_sub(rownames(tr_data_cov),1,4)
  df <- na.omit(df)
  ## do scatter plot 
  p <- ggplot(df, aes(x = x, y = y,label = smp)) +
    geom_point(aes(colour=status,alpha = 0.5),
               position = position_jitter(width = 0.02, height = 0.02)) +
    geom_text_repel(aes(label = smp))+
    labs(x = paste("x_", col), y = paste("y_", col), title = paste("scatter.plot -", col))+
    theme_classic()
    print(p)
    ## wilcox test
    ##usong wilcox function
    # wilcox_test_1 <-  compare_means(y~status,df,paired = F ,method = "wilcox.test",p.adjust.method = "BH")
    wilcox_test <- wilcox.test(df$y[df$status=="GP"],
                               df$y[df$status=="GC"],paired=F)
    
    p_value<- wilcox_test$p.value
    pvalue_list[[col]] <- p_value
    
    ##排除检验结果为空的情况
    
  if(!is.na(p_value)&&p_value< 0.05&&length(p_value)!=0){ 
        p1 <- ggplot(df, aes(x = status, y = y)) +
              geom_violin(aes(fill=status,alpha=1),
              trim = FALSE, lwd = 0.8, width = 1.2,color="white") +
              scale_fill_manual(values =c("#FC9595", "#C4E3FF"))+
              geom_boxplot(aes(fill=status),
                            outlier.shape = NA,lwd=0.4,width=0.03)+
              scale_fill_manual(values =c("#FB4F4F", "#93CDFF"),guide = 'none')+
              labs(y ="Methylation level", x = paste("y_", col), title = paste("region -", col))+
              geom_signif(comparisons = list(c("GP", "GC")),
                          textsize = 8,annotations = c(paste0("p = ", round(p_value, 10))),
                          y_position = 1.2,vjust = -0.5)+
              theme_classic()+
              theme(axis.text.x = element_text(size = 15),
                    axis.text.y = element_text(size = 15),
                    axis.text = element_text(size = 15,colour = "black"),
                    axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.title = element_text(size = 15,colour = "black"),
                    legend.position="none")
        # ggsave(plot = p1,filename =paste0(fig_proj,"fig4C_gene_violin_plot_20240829/","meth_",
        #                                   col,"_violin_plot_20240829.pdf"),width = 6,height = 4,dpi = 300)

        print(p1)
         }else{
             print(paste0(col,"_wilcox p_value less than 0.05,and the p is ",p_value))
            count=count+1 }
          print(count)
 }

## multi comparision  adjust 
p_DMR <- data.frame(
                    gene_region=names(pvalue_list),
                    p_value=unlist(pvalue_list)
)

p_DMR$p_adjust <- p.adjust(pvalue_list,method="BH")%>%unlist()

df <- p_DMR

meth_DMR_pvalue <- merge(df,merge_smp_meth_ann,by="gene_region")

# write.csv(meth_DMR_pvalue,
#           "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/meth_DMR_p_value_202420909.csv",
#           col.names = F,row.names = F,quote = F)

meth_DMR_pvalue <- read.csv("/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/meth_DMR_p_value_202420909.csv")

DMR_647_region_padjust <- meth_DMR_pvalue[,c(1,3)]

df <- meth_DMR_pvalue

df <- df %>%
  separate(gene_region, into = c("chr","start","end","pvalue","gene"), sep = "_")


# write.csv(df, "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/DMR_regions_padjust_202420909.csv",col.names = F,row.names = F,quote = F)

# 查看结果
print(df)
```
## fig4C (20240902 modified) associated gene at dmr regions between GC and GP 
### do vilolinplot for all associate genes
```{r}
## import data
sample_list <- c("gp01_L14-1550","gp03_L15-31","gp04_L15-12","gp05_L15-1550",
                  "gp06_L15-13","gp07_L19-5","gp08_L19-6","gp09_L19-7","gp10_L19-8","gp11_L11-1",
                  "gp12_L11-3","gp14_L11-8","gp15_L12-1","gp17_L16-2","gp18_L16-4",
                  "gp19_L16-7","gp20_L16-8","sc01_14-Z","sc02_L14-1502","sc04_L14-24",
                  "sc05_L15-24","sc06_L15-33","sc07_L15-25","sc08_L15-1608","sc09_L19-1","sc10_L19-2","sc11_L19-3",
                  "sc12_L19-4","sc13_L11-2","sc14_L11-5","sc15_L12-2","sc16_L12-3","sc17_L12-4","sc18_L12-5",
                  "sc19_L12-6","sc20_L12-8","sc21_L16-1","sc23_L16-6")


dat_proj <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/filter_sample/dmr_file_delta0.15/"
## change data format
sample_DMR_meth <- read.table(paste0(dat_proj,"dmrs_region_list.bed"),col.names = "chr_region")
sample_DMR_cov <- read.table(paste0(dat_proj,"dmrs_region_list.bed"),col.names = "chr_region")

for (i_sample in sample_list){
  a <- read.table(paste0(dat_proj,i_sample,"_dmr_meth.bed"),col.name = c("chromosome","start","end","mod","unmod","meth_level"))

 sample_DMR_meth[[i_sample]] <- a$meth_level
 sample_DMR_cov[[i_sample]] <- a$mod+a$unmod
}

## import dmr_region genes
dmr_ann <- read.csv("/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/meth_merge_change.csv")
## change data format
dmr_ann[, c(1:5)] <- lapply(dmr_ann[,c(1:5)], as.character)
dmr_ann$gene_region <-  apply(dmr_ann[, c(1:5)], 1, function(x) paste(x, collapse = "_"))

dmr_ann$chr_region <- apply(dmr_ann[, c(1:3)], 1, function(x) paste(x, collapse = "_"))

dmr_ann_df <- dmr_ann[,c(45,44,6:43)]
## merge smp_meth_data into gene annotation
merge_smp_meth_ann <- merge(dmr_ann_df,sample_DMR_meth,by="chr_region")
smp_meth_ann <- merge_smp_meth_ann[,c(2,41:78)]

##merge smp_cov_data into gene annotation
merge_smp_cov_ann <- merge(dmr_ann_df,sample_DMR_cov,by="chr_region")
smp_cov_ann <- merge_smp_cov_ann[,c(2,41:78)]

## sample_meth
data <- smp_meth_ann
rownames(data) <- as.character(data$gene_region)
data <- data[, -1]
tr_data <- t(data)
rownames(tr_data) <- colnames(data)
colnames(tr_data) <- rownames(data)
tr_data_meth <- as.data.frame(tr_data)
# tr_data_meth$smp_type <- meta_df$sc_type
## sample_cov
data_cov <- smp_cov_ann
rownames(data_cov) <- as.character(data_cov$gene_region)
data_cov <- data_cov[, -1]
tr_data_co <- t(data_cov)
rownames(tr_data_co) <- colnames(data_cov)
colnames(tr_data_co) <- rownames(data_cov)
tr_data_cov<- as.data.frame(tr_data_co)
# tr_data_cov$smp_type <- meta_df$sc_type

## write a loop to do plot 

common_cols <- intersect(names(tr_data_cov), names(tr_data_meth))

count=0
p_adjust <- list()
p_value <- list()
## 
for (col in common_cols) {
  df <- data.frame(x = tr_data_cov[[col]], y = tr_data_meth[[col]]) 
  # df$status <- str_sub(rownames(tr_data_cov),1,2)
  # df$smp <- str_sub(rownames(tr_data_cov),1,4)
  df$status <- meta_df$sc_type%>%as.factor()
  df$smp <- str_sub(rownames(tr_data_cov),1,4)
  df <- na.omit(df)
  ## do scatter plot 
  p <- ggplot(df, aes(x = x, y = y,label = smp)) +
    geom_point(aes(colour=status,alpha = 0.5),
               position = position_jitter(width = 0.02, height = 0.02)) +
    geom_text_repel(aes(label = smp))+
    labs(x = paste("x_", col), y = paste("y_", col), title = paste("scatter.plot -", col))+
    theme_classic()
    print(p)
    ## wilcox test
    ##usong wilcox function
    # wilcox_test_1 <-  compare_means(y~status,df,paired = F ,method = "wilcox.test",p.adjust.method = "BH")
    wilcox_test <- wilcox.test(df$y[df$status=="GP"],
                               df$y[df$status=="GC"],paired=F)
    
    p_value<- wilcox_test$p.value
    p_adjust[[col]] <- p_value
    
    ##排除检验结果为空的情况
    
  if(!is.na(p_value)&&p_value< 0.05&&length(p_value)!=0){ 
        p1 <- ggplot(df, aes(x = status, y = y)) +
              geom_violin(aes(fill=status,alpha=1),scale = "width") +
              scale_fill_manual(values =c("#FC9595", "#C4E3FF"))+
              geom_boxplot(aes(fill=status),
                            outlier.shape = NA,lwd=0.4,width=0.05)+
              scale_fill_manual(values =c("#FB4F4F", "#93CDFF"),guide = 'none')+
              labs(y ="Methylation level", x = paste("y_", col), title = paste0("p = ", round(p_value, 10)))+
              # geom_signif(comparisons = list(c("GP", "GC")),
              #             textsize = 8,annotations = c(paste0("p = ", round(p_value, 10))),
              #             y_position = 1.2,vjust = -0.5)+
              theme_classic()+
              theme(axis.text.x = element_text(size = 15),
                    axis.text.y = element_text(size = 15),
                    axis.text = element_text(size = 15,colour = "black"),
                    axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.title = element_text(size = 15,colour = "black"),
                    legend.position="none")
        ggsave(plot = p1,filename =paste0(fig_proj,"fig4C_gene_violin_plot_20240902/","meth_",
                col,"_violin_plot_20240902.pdf"),width = 6,height = 4,dpi = 300)


        
    #     p <- ggplot(df_1, aes(x = smp_type, y = probility)) +
    #     geom_violin(aes(fill=smp_type,alpha=1),scale = "width") +
    #     scale_fill_manual(values =c("#FC9595", "#C4E3FF"))+
    #     geom_boxplot(#aes(fill=smp_type),
    #             width=0.15)+
    #   #scale_fill_manual(values =c("#FB4F4F", "#93CDFF"),guide = 'none')+
    #   labs(y =paste0("Probability of originating"), x =" ", 
    #    title = paste0("cell type ", i, " p = ", round(p.val, 3) ,""))+
    #   theme_classic()+ 
    # theme(
    #         axis.text.x = element_text(size = 15),
    #         axis.text.y = element_text(size = 15),
    #         axis.text = element_text(size = 15,colour = "black"),
    #         axis.title = element_text(size = 15,colour = "black"),
    #         legend.position="none")
        
        
        
        
        
        
        
        
        
        
        # ggsave(plot = p1,filename =paste0(fig_proj,"fig4C_gene_violin_plot_20240829/","meth_",
                                          # col,"_violin_plot_20240829.pdf"),width = 6,height = 4,dpi = 300)

        print(p1)
         }else{
             print(paste0(col,"_wilcox p_value less than 0.05,and the p is ",p_value))
            count=count+1 }
          print(count)
 }



```
##fig4A GO analysis barplot about DMR region between GC and GP

```{r,fig.width=6,fig.height=6}
## import data
fig4_dat <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/fig4_GO_data/"
smp_go_table <- read_excel(paste0(fig4_dat,"GO_Biological_Process_2023_table_filter_dmr (2).xlsx"))
smp_go_top10_df <- smp_go_table[1:11,c(1,3,4)]
smp_go_top10_df$log10p_adjust <- -log10(smp_go_top10_df$`Adjusted P-value`)
smp_go_top10_df$Term <- factor(smp_go_top10_df$Term, levels = rev(smp_go_top10_df$Term))
##do barplot for top 10 GO term
p1 <- ggplot(smp_go_top10_df,aes(y=Term,x =log10p_adjust))+
      geom_bar(stat="identity",color="black",fill="#62A4CE",width = 0.5)+
      labs(x="-log10P",y=" ")+
      xlim(0,2.5)+
      theme_classic()+
      theme(axis.text.x = element_text(size = 15),
                    axis.text.y = element_text(size = 15),
                    axis.text = element_text(size = 15,colour = "black"),
                    axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.title = element_text(size = 15,colour = "black"))
# ggsave(plot = p1,filename = paste0(fig_proj,"fig4A_filter_dmr_smp_GO_top10_p.pdf"),width = 12,height = 4,dpi = 300)
print(p1)
```



##fig4A GO analysis draw top10 GO terms barplot code shared by professor Hu 
```{r,fig.width=20,fig.height=8}
# Function to generate a plot from a given file
# generate_plot <- function(file_path) {
#   data <- read.delim(file_path, header = TRUE, sep = "\t")
  data <- smp_go_table[1:11,]


  data <- data %>%
    mutate(
      Combined_score = data$`Combined Score`,
      EnrichmentFDR = -log10(data$`Adjusted P-value`),
      N_genes = sapply(strsplit(as.character(Genes), ";"), length)
      )

  data <- data[data$EnrichmentFDR > 1.30103 ,]
  
  plot <- ggplot(data, aes(x = reorder(Term, Combined_score), y = Combined_score)) +
    geom_point(aes(size = N_genes, color = EnrichmentFDR)) +
    scale_color_gradient(low = "blue", high = "red") +
    coord_flip() +
    labs(
      title = "Top10 GO term",
      x = "",
      y = "Combined score",
      color = "EnrichmentFDR",
      size = "Genes"
    ) +
    theme_minimal() +
   theme(axis.text.x = element_text(size = 15),
                    axis.text.y = element_text(size = 15),
                    axis.text = element_text(size = 15,colour = "black"),
                    axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.title = element_text(size = 15,colour = "black"),
                    legend.title = element_text(size = 15),
                    legend.text = element_text(size = 15))
  
  plot
# ggsave(plot = plot,filename = paste0(fig_proj,"fig4A_filter_dmr_smp_GO_top10_p_20240722.pdf"),
#          device = "pdf",width = 15, height = 6)
  
```



## fig4B  Heatmap for Top 10 gene terms 
```{r,fig.width=8,fig.height=8}
## import data
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/"
go_gene <- read.csv(paste0(dat_dir,"filter_dmr_top10_gene.csv"))
## modified 20240720 import duplicate gene
duplicate_gene <- read.csv(paste0(dat_dir,"top10_GO_term_duplicate_genes.csv"))
unique_genes <- unique(duplicate_gene$Gene)


# go_gene[,7:44] <- as.numeric(go_gene[,7:44])

# go_gene[, 7:44] <- lapply(go_gene[, 7:44], function(x) as.numeric(as.character(x)))

## 
go_gene_p <- go_gene[go_gene$p_adjust<0.05,]

df_B_sorted <- go_gene_p %>% 
  filter(Genes %in% unique_genes) %>% 
  arrange(factor(Genes, levels = unique_genes))





heatmap_dat <- go_gene_p[,3:45]
rownames(heatmap_dat) <- heatmap_dat$Genes
gp_mat <- heatmap_dat[,6:22]%>%as.matrix()
rownames(gp_mat) <- rownames(heatmap_dat)
gc_mat <- heatmap_dat[,23:43]%>%as.matrix()
rownames(gc_mat) <- rownames(heatmap_dat)
## annotation setting 


type_colors <- c("GC" = "#FB8D8D", "GP" = "#C0E1FF")
status_colors <- c("Control" = "#63A6D0", "Early Stage" = "#F9A68A", "Advanced Stage" = "#D34447")

gender_colors <- c("Male" = "#AACBFE", "Female" = "#DE8686")

my_color=c("#4675FE","#F9F5C9","#FCA605")
# type_colors=list(Type=c(GP="#C0E1FF",GC="#FB8D8D"))
# 
# gender_color=list(Gender=c(Male="#AACBFE",Female="#DE8686"))
# 
# stage_color=list(Stage=c("Control"="#63A6D0","Early Stage"="#F9A68A","Advanced Stage"="#D34447"))

ann_ht1 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[1:17],
                             
                             Gender=meta_df$sc_gender[1:17],
                              Type=meta_df$sc_type[1:17],
                             annotation_name_side = "left",
                               col=list(
                                 Type = type_colors,
                                  Status = status_colors,
                                  Gender = gender_colors))

ann_ht2 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[18:38],
                            Gender=meta_df$sc_gender[18:38],
                            Type=meta_df$sc_type[18:38],
                            col=list(Type = type_colors,
                                      Status = status_colors,
                                      Gender = gender_colors),
                                      annotation_name_side = "right")

## do heatmap
# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_top10_gene_heatmap.pdf"),width = 8,height = 10)
heat_map_1 <- Heatmap(gp_mat,name="gp",
                      cluster_rows = F,
                      cluster_columns = T,
                      top_annotation = ann_ht1,
                      show_column_names = F,
                       col = colorRampPalette(my_color)(100),
                       na_col = "white",width = 6,height = 8
                      )
heat_map_2 <- Heatmap(gc_mat,name="gc",
                      cluster_rows = F,
                      cluster_columns = T,
                      top_annotation = ann_ht2,
                      show_column_names = F,
                      col = colorRampPalette(my_color)(100),
                      na_col = "white",width = 6,height = 8,
                      row_dend_side = "right")

heat_map_1+heat_map_2
# dev.off()

```

## fig4B  Heatmap for intresting GO 4  gene terms 
```{r,fig.width=15,fig.height=10}
## import data
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/"
go_gene_4 <- read.csv(paste0(dat_dir,"Go4_genes_meth.csv"))
## modified 20240720 import duplicate gene
# duplicate_gene <- read.csv(paste0(dat_dir,"top10_GO_term_duplicate_genes.csv"))
# unique_genes <- unique(duplicate_gene$Gene)


# go_gene[,7:44] <- as.numeric(go_gene[,7:44])

# go_gene[, 7:44] <- lapply(go_gene[, 7:44], function(x) as.numeric(as.character(x)))

## 
go4_gene_p <- go_gene_4[go_gene_4$p_adjust<0.05,]

# df_B_sorted <- go_gene_p %>% 
#   filter(Genes %in% unique_genes) %>% 
#   arrange(factor(Genes, levels = unique_genes))

##sort datafram arrange genes
order_name <- c( "DPYSL3","RAPGEF2", "NLGN1", #duplicate gene
                  "TENM3" ,"MAGI2","ROBO1","SLIT2","EPHA3","DDR2","NRXN1","CDH12",  
                  "SLITRK6","FN1","PKP3","CDH9","POU4F1","CDH8","ADGRL3","SMARCB1","ARID1B","SMARCA2","KANK1", 
                  "DCC","LPAR1" )

go_gene_4$Genes <- factor(go_gene_4$Genes, levels = order_name)

sorted_go_gene_4 <- go_gene_4 %>% arrange(Genes)
##do heatmap 
heatmap_dat <- sorted_go_gene_4[,c(2,10:47)]
rownames(heatmap_dat) <- heatmap_dat$Genes
gp_mat <- heatmap_dat[,2:18]%>%as.matrix()
rownames(gp_mat) <- rownames(heatmap_dat)
gc_mat <- heatmap_dat[,19:39]%>%as.matrix()
rownames(gc_mat) <- rownames(heatmap_dat)
## annotation setting 


type_colors <- c("GC" = "#FB8D8D", "GP" = "#C0E1FF")
status_colors <- c("Control" = "#63A6D0", "Early Stage" = "#F9A68A", "Advanced Stage" = "#D34447")

gender_colors <- c("Male" = "#AACBFE", "Female" = "#DE8686")

my_color=c("#4675FE","#F9F5C9","#FCA605")
# type_colors=list(Type=c(GP="#C0E1FF",GC="#FB8D8D"))
# 
# gender_color=list(Gender=c(Male="#AACBFE",Female="#DE8686"))
# 
# stage_color=list(Stage=c("Control"="#63A6D0","Early Stage"="#F9A68A","Advanced Stage"="#D34447"))

row_groups <- c(rep("GO:0031346;GO:0031345", 2),
                rep("GO:0031346;GO:0034329",1),
                rep("GO:0031346",6),
                rep("GO:0034329",9),
                rep("GO:0045582",3),
               rep("GO:0031345",3))

## select annotated DMR region CHD12 enhancer  
row_enhancer <- c(rep("No",10),"Yes",rep("No",12),"Yes")%>%as.factor()

# row_ann <- rowAnnotation(GO=row_groups,col =list(GO=row_group_col),labels=GO_labels)

ann_ht1 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[1:17],
                             
                             Gender=meta_df$sc_gender[1:17],
                              Type=meta_df$sc_type[1:17],
                             annotation_name_side = "left",
                               col=list(
                                 Type = type_colors,
                                  Status = status_colors,
                                  Gender = gender_colors))

ann_ht2 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[18:38],
                            Gender=meta_df$sc_gender[18:38],
                            Type=meta_df$sc_type[18:38],
                            col=list(Type = type_colors,
                                      Status = status_colors,
                                      Gender = gender_colors),
                                      annotation_name_side = "right")



# row_annots <- rowAnnotation(GO = row_groups,show_annotation_name = TRUE,
#                             col = list(GO = c("GO:0034329"="#2F7EBC",
#                                              
#                                              "GO:0031345"="#F96949",
#                                              
#                                              "GO:0031346"="#50B8DD",
#                                              
#                                               "GO:0045582"="#FCBC85",
#                                              
#                                              "GO:0031346;GO:0034329"="#77C8A6",
#                                              
#                                              "GO:0031346;GO:0031345"="#FCEBDE"
#                                             )),
# labels=c("Cell Junction Assembly",
#           "Positive Regulation Of T Cell Differentiation",
#          "Positive Regulation Of Cell Projection Organization",
#         "Negative Regulation Of Cell Projection Organization",
#         "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly",
#         "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization"))
#                  
# annotation_legend_param = list(GO = list(title = " "))
# labels=c("GO:0031346","GO:0034329","GO:0045582","GO:0031345","GO:0031346;GO:0031345","GO:0031346;GO:0034329")

row_enhacer_col <- rowAnnotation(Enhancer=row_enhancer,col=list(Enhancer=c("No"="#999999","Yes"="#F2F2F2")))


row_annots <- rowAnnotation(GO = row_groups,
                           col = list(GO = c("GO:0034329"="#2F7EBC",

                                             "GO:0031345"="#F96949",

                                             "GO:0031346"="#50B8DD",

                                              "GO:0045582"="#FCBC85",

                                             "GO:0031346;GO:0034329"="#77C8A6",

                                             "GO:0031346;GO:0031345"="#FCEBDE"
                                            )),
                           annotation_legend_param = list(
                             GO = list(
                               at = c("GO:0034329", "GO:0031345", "GO:0031346", "GO:0045582", 
                                      "GO:0031346;GO:0034329", "GO:0031346;GO:0031345"),
                               labels = c("Cell Junction Assembly", 
                                          "Positive Regulation Of T Cell Differentiation", 
                                          "Positive Regulation Of Cell Projection Organization", 
                                          "Negative Regulation Of Cell Projection Organization", 
                                          "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly", 
                                          "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization")
                             )
                           ))









## do heatmap
# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_top10_gene_heatmap_20240731.pdf"),width = 16,height = 10)
heat_map_1 <- Heatmap(gp_mat,name="gp",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                     # right_annotation = row_annots,
                      top_annotation = ann_ht1,
                      show_column_names = F,
                     show_row_dend = F,
                     show_row_names = T,
                     row_title = NULL,
                       col = colorRampPalette(my_color)(100),
                       na_col = "white",width =unit(4,"cm"),
                     height = unit(10,"cm")
                      )
heat_map_2 <- Heatmap(gc_mat,name="gc",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                      right_annotation =c(row_annots,row_enhacer_col) ,
                      top_annotation = ann_ht2,
                      show_column_names = F,
                      show_row_dend = F,
                      show_row_names = T,
                       row_title = NULL,
                      col = colorRampPalette(my_color)(100),
                      na_col = "white",width = unit(4,"cm"),
                      height = unit(10,"cm"),
                      row_dend_side = "right")


p <- heat_map_1+heat_map_2

# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_top10_gene_heatmap_20240801.pdf"),width = 16,height = 10)
draw(p, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")


# dev.off()

```

## fig4B  Add DMRS region for the annotated genes .Heatmap of  intrested GO 4  gene terms(20240829 modified)
```{r,fig.width=15,fig.height=10}
## import data
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/"
go_gene_4 <- read.csv(paste0(dat_dir,"Go4_genes_meth.csv"))
## modified 20240720 import duplicate gene
# duplicate_gene <- read.csv(paste0(dat_dir,"top10_GO_term_duplicate_genes.csv"))
# unique_genes <- unique(duplicate_gene$Gene)


# go_gene[,7:44] <- as.numeric(go_gene[,7:44])

# go_gene[, 7:44] <- lapply(go_gene[, 7:44], function(x) as.numeric(as.character(x)))

## 
go4_gene_p <- go_gene_4[go_gene_4$p_adjust<0.05,]

# df_B_sorted <- go_gene_p %>% 
#   filter(Genes %in% unique_genes) %>% 
#   arrange(factor(Genes, levels = unique_genes))

##sort datafram arrange genes
order_name <- c( "DPYSL3","RAPGEF2", 
                 "NLGN1", #duplicate gene
                  "TENM3" ,"MAGI2","ROBO1","SLIT2","EPHA3","DDR2",
                 "NRXN1","CDH12","SLITRK6","FN1","PKP3","CDH9","POU4F1","CDH8","ADGRL3",
                 "SMARCB1","ARID1B","SMARCA2",
                  "KANK1","DCC","LPAR1" )

go_gene_4$Genes <- factor(go_gene_4$Genes, levels = order_name)

sorted_go_gene_4 <- go_gene_4 %>% arrange(Genes)
##add new cols contain gene_chr_start_end information 

sorted_go_gene_4[,4:8] <- lapply(sorted_go_gene_4[,4:8],as.character)
sorted_go_gene_4$new_col <- apply(sorted_go_gene_4[,c(2,5,4,6)],1,function(x) paste(x,collapse = "_"))

## extract the data do heatmap

heatmap_dat <- sorted_go_gene_4[,10:48]
rownames(heatmap_dat) <- heatmap_dat$new_col

gp_mat <- heatmap_dat[,1:17]%>%as.matrix()
rownames(gp_mat) <- rownames(heatmap_dat)
gc_mat <- heatmap_dat[,18:38]%>%as.matrix()
rownames(gc_mat) <- rownames(heatmap_dat)
## annotation setting 


type_colors <- c("GC" = "#FB8D8D", "GP" = "#C0E1FF")
status_colors <- c("Control" = "#63A6D0", "Early Stage" = "#F9A68A", "Advanced Stage" = "#D34447")

gender_colors <- c("Male" = "#AACBFE", "Female" = "#DE8686")

my_color=c("#4675FE","#F9F5C9","#FCA605")
# type_colors=list(Type=c(GP="#C0E1FF",GC="#FB8D8D"))
# 
# gender_color=list(Gender=c(Male="#AACBFE",Female="#DE8686"))
# 
# stage_color=list(Stage=c("Control"="#63A6D0","Early Stage"="#F9A68A","Advanced Stage"="#D34447"))

row_groups <- c(rep("GO:0031346;GO:0031345", 2),
                rep("GO:0031346;GO:0034329",1),
                rep("GO:0031346",6),
                rep("GO:0034329",9),
                rep("GO:0045582",3),
               rep("GO:0031345",3))

## select annotated DMR region CHD12 enhancer  
row_enhancer <- c(rep("No",10),"Yes",rep("No",12),"Yes")%>%as.factor()

# row_ann <- rowAnnotation(GO=row_groups,col =list(GO=row_group_col),labels=GO_labels)

ann_ht1 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[1:17],
                             
                             Gender=meta_df$sc_gender[1:17],
                              Type=meta_df$sc_type[1:17],
                             annotation_name_side = "left",
                               col=list(
                                 Type = type_colors,
                                  Status = status_colors,
                                  Gender = gender_colors))

ann_ht2 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[18:38],
                            Gender=meta_df$sc_gender[18:38],
                            Type=meta_df$sc_type[18:38],
                            col=list(Type = type_colors,
                                      Status = status_colors,
                                      Gender = gender_colors),
                                      annotation_name_side = "right")



# row_annots <- rowAnnotation(GO = row_groups,show_annotation_name = TRUE,
#                             col = list(GO = c("GO:0034329"="#2F7EBC",
#                                              
#                                              "GO:0031345"="#F96949",
#                                              
#                                              "GO:0031346"="#50B8DD",
#                                              
#                                               "GO:0045582"="#FCBC85",
#                                              
#                                              "GO:0031346;GO:0034329"="#77C8A6",
#                                              
#                                              "GO:0031346;GO:0031345"="#FCEBDE"
#                                             )),
# labels=c("Cell Junction Assembly",
#           "Positive Regulation Of T Cell Differentiation",
#          "Positive Regulation Of Cell Projection Organization",
#         "Negative Regulation Of Cell Projection Organization",
#         "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly",
#         "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization"))
#                  
# annotation_legend_param = list(GO = list(title = " "))
# labels=c("GO:0031346","GO:0034329","GO:0045582","GO:0031345","GO:0031346;GO:0031345","GO:0031346;GO:0034329")

row_enhacer_col <- rowAnnotation(Enhancer=row_enhancer,col=list(Enhancer=c("No"="#999999","Yes"="#F2F2F2")))


row_annots <- rowAnnotation(GO = row_groups,
                           col = list(GO = c("GO:0034329"="#2F7EBC",

                                             "GO:0031345"="#F96949",

                                             "GO:0031346"="#50B8DD",

                                              "GO:0045582"="#FCBC85",

                                             "GO:0031346;GO:0034329"="#77C8A6",

                                             "GO:0031346;GO:0031345"="#FCEBDE"
                                            )),
                           annotation_legend_param = list(
                             GO = list(
                               at = c("GO:0034329", "GO:0031345", "GO:0031346", "GO:0045582", 
                                      "GO:0031346;GO:0034329", "GO:0031346;GO:0031345"),
                               labels = c("Cell Junction Assembly", 
                                          "Positive Regulation Of T Cell Differentiation", 
                                          "Positive Regulation Of Cell Projection Organization", 
                                          "Negative Regulation Of Cell Projection Organization", 
                                          "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly", 
                                          "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization")
                             )
                           ))









## do heatmap
# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_top10_gene_heatmap_20240731.pdf"),width = 16,height = 10)
heat_map_1 <- Heatmap(gp_mat,name="gp",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                     # right_annotation = row_annots,
                      top_annotation = ann_ht1,
                      show_column_names = F,
                     show_row_dend = F,
                     show_row_names = T,
                     row_title = NULL,
                       col = colorRampPalette(my_color)(100),
                       na_col = "white",width =unit(4,"cm"),
                     height = unit(10,"cm")
                      )
heat_map_2 <- Heatmap(gc_mat,name="gc",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                      right_annotation =c(row_annots,row_enhacer_col) ,
                      top_annotation = ann_ht2,
                      show_column_names = F,
                      show_row_dend = F,
                      show_row_names = T,
                       row_title = NULL,
                      col = colorRampPalette(my_color)(100),
                      na_col = "white",width = unit(4,"cm"),
                      height = unit(10,"cm"),
                      row_dend_side = "right")


p <- heat_map_1+heat_map_2

# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_spe4_gene_heatmap_20240829.pdf"),width = 16,height = 10)
draw(p, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")


# dev.off()

```
## fig4B  Add DMRS region for the annotated genes .Heatmap of  intrested GO 4  gene terms(20240909 modified)
```{r,fig.width=15,fig.height=10}
## import data
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/"
go_gene_4 <- read.csv(paste0(dat_dir,"Go4_genes_meth_20240909.csv"))
## modified 20240720 import duplicate gene
# duplicate_gene <- read.csv(paste0(dat_dir,"top10_GO_term_duplicate_genes.csv"))
# unique_genes <- unique(duplicate_gene$Gene)


# go_gene[,7:44] <- as.numeric(go_gene[,7:44])

# go_gene[, 7:44] <- lapply(go_gene[, 7:44], function(x) as.numeric(as.character(x)))

## 
go4_gene_p <- go_gene_4[go_gene_4$P_Adjust<0.05,]

# df_B_sorted <- go_gene_p %>% 
#   filter(Genes %in% unique_genes) %>% 
#   arrange(factor(Genes, levels = unique_genes))

##sort datafram arrange genes
order_name <- c( "DPYSL3","RAPGEF2",
                 "NLGN1", #duplicate gene
                  "TENM3" ,"MAGI2","SLIT2",
                 "NRXN1","CDH12","FN1","CDH8",
                 "SMARCB1","SMARCA2",
                 "KANK1")

go_gene_4$Genes <- factor(go_gene_4$Genes, levels = order_name)

sorted_go_gene_4 <- go_gene_4 %>% arrange(Genes)
##add new cols contain gene_chr_start_end information 

sorted_go_gene_4[,2:6] <- lapply(sorted_go_gene_4[,2:6],as.character)
sorted_go_gene_4$new_col <- apply(sorted_go_gene_4[,c(2,4,5,6)],1,function(x) paste(x,collapse = "_"))

## extract the data do heatmap

heatmap_dat <- sorted_go_gene_4[,8:46]
rownames(heatmap_dat) <- heatmap_dat$new_col

gp_mat <- heatmap_dat[,1:17]%>%as.matrix()

rownames(gp_mat) <- rownames(heatmap_dat)

gc_mat <- heatmap_dat[,18:38]%>%as.matrix()

rownames(gc_mat) <- rownames(heatmap_dat)
## annotation setting 


type_colors <- c("GC" = "#FB8D8D", "GP" = "#C0E1FF")
status_colors <- c("Control" = "#63A6D0", "Early Stage" = "#F9A68A", "Advanced Stage" = "#D34447")

gender_colors <- c("Male" = "#AACBFE", "Female" = "#DE8686")

my_color=c("#4675FE","#F9F5C9","#FCA605")
# type_colors=list(Type=c(GP="#C0E1FF",GC="#FB8D8D"))
# 
# gender_color=list(Gender=c(Male="#AACBFE",Female="#DE8686"))
# 
# stage_color=list(Stage=c("Control"="#63A6D0","Early Stage"="#F9A68A","Advanced Stage"="#D34447"))

row_groups <- c(rep("GO:0031346;GO:0031345", 2),
                rep("GO:0031346;GO:0034329",1),
                rep("GO:0031346",3),
                rep("GO:0034329",4),
                rep("GO:0045582",2),
               rep("GO:0031345",1))

## select annotated DMR region CHD12 enhancer  
row_enhancer <- c(rep("No",7),"Yes",rep("No",5))%>%as.factor()

# row_ann <- rowAnnotation(GO=row_groups,col =list(GO=row_group_col),labels=GO_labels)

ann_ht1 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[1:17],
                             Gender=meta_df$sc_gender[1:17],
                              Type=meta_df$sc_type[1:17],
                             annotation_name_side = "left",
                               col=list(
                                 Type = type_colors,
                                  Status = status_colors,
                                  Gender = gender_colors))

ann_ht2 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[18:38],
                            Gender=meta_df$sc_gender[18:38],
                            Type=meta_df$sc_type[18:38],
                            col=list(Type = type_colors,
                                      Status = status_colors,
                                      Gender = gender_colors),
                                      annotation_name_side = "right")



# row_annots <- rowAnnotation(GO = row_groups,show_annotation_name = TRUE,
#                             col = list(GO = c("GO:0034329"="#2F7EBC",
#                                              
#                                              "GO:0031345"="#F96949",
#                                              
#                                              "GO:0031346"="#50B8DD",
#                                              
#                                               "GO:0045582"="#FCBC85",
#                                              
#                                              "GO:0031346;GO:0034329"="#77C8A6",
#                                              
#                                              "GO:0031346;GO:0031345"="#FCEBDE"
#                                             )),
# labels=c("Cell Junction Assembly",
#           "Positive Regulation Of T Cell Differentiation",
#          "Positive Regulation Of Cell Projection Organization",
#         "Negative Regulation Of Cell Projection Organization",
#         "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly",
#         "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization"))
#                  
# annotation_legend_param = list(GO = list(title = " "))
# labels=c("GO:0031346","GO:0034329","GO:0045582","GO:0031345","GO:0031346;GO:0031345","GO:0031346;GO:0034329")

row_enhacer_col <- rowAnnotation(Enhancer=row_enhancer,col=list(Enhancer=c("No"="#999999","Yes"="#F2F2F2")))


row_annots <- rowAnnotation(GO = row_groups,
                           col = list(GO = c("GO:0034329"="#2F7EBC",

                                             "GO:0031345"="#F96949",

                                             "GO:0031346"="#50B8DD",

                                              "GO:0045582"="#FCBC85",

                                             "GO:0031346;GO:0034329"="#77C8A6",

                                             "GO:0031346;GO:0031345"="#FCEBDE"
                                            )),
                           annotation_legend_param = list(
                             GO = list(
                               at = c("GO:0034329", "GO:0031345", "GO:0031346", "GO:0045582", 
                                      "GO:0031346;GO:0034329", "GO:0031346;GO:0031345"),
                               labels = c("Cell Junction Assembly", 
                                          "Positive Regulation Of T Cell Differentiation", 
                                          "Positive Regulation Of Cell Projection Organization", 
                                          "Negative Regulation Of Cell Projection Organization", 
                                          "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly", 
                                          "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization")
                             )
                           ))









## do heatmap
# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_top10_gene_heatmap_20240731.pdf"),width = 16,height = 10)
heat_map_1 <- Heatmap(gp_mat,name="gp",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                     # right_annotation = row_annots,
                      top_annotation = ann_ht1,
                      show_column_names = F,
                     show_row_dend = F,
                     show_row_names = T,
                     row_title = NULL,
                       col = colorRampPalette(my_color)(100),
                       na_col = "white",width =unit(4,"cm"),
                     height = unit(10,"cm")
                      )
heat_map_2 <- Heatmap(gc_mat,name="gc",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                      right_annotation =c(row_annots,row_enhacer_col) ,
                      top_annotation = ann_ht2,
                      show_column_names = F,
                      show_row_dend = F,
                      show_row_names = T,
                       row_title = NULL,
                      col = colorRampPalette(my_color)(100),
                      na_col = "white",width = unit(4,"cm"),
                      height = unit(10,"cm"),
                      row_dend_side = "right")


p <- heat_map_1+heat_map_2

# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_spe4_gene_heatmap_20240909.pdf"),width = 16,height = 10)
# draw(p, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")
# 
# 
# dev.off()

```



## technical
```{r}
sessionInfo()
```

