---
title: "fig4_plot(filter dmr remove blacklist)"
author: "Zhenna Jiao"
date: "2024-03-04(last modified: `r Sys.Date()`)"
output: 
    prettydoc::html_pretty:
    theme: cayman
    highlight: github
---
## Load packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggbiplot)
library(ggplot2)
library(corrplot)
library(magrittr)
library(RColorBrewer)
library(stringr)
library(ggpubr)
library(ggrepel)
library(readxl)
library(ComplexHeatmap)
library(scales)
library(dplyr)
```


## Set some dir
```{r}
## smp_meta_dat_df
meta_df <- readRDS("/home/jiaozhenna/methylation/project/EM_seq/data/metadata/meta_df.rds")
## fig dir
meta_scstage_df <- readRDS("/home/jiaozhenna/methylation/project/EM_seq/data/metadata/meta_scstage_df.rds")
fig_proj <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/figures/fig4/"
sc_dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/filter_sample/scstage_dmr_change/"
```

##fig4 a GO analysis draw top10 GO terms barplot  
```{r,fig.width=20,fig.height=8}
## import data
fig4_dat <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/fig4_GO_data/"
smp_go_table <- read_excel(paste0(fig4_dat,"GO_Biological_Process_2023_table_filter_dmr (2).xlsx"))
smp_go_top10_df <- smp_go_table[1:11,c(1,3,4)]
smp_go_top10_df$log10p_adjust <- -log10(smp_go_top10_df$`Adjusted P-value`)
smp_go_top10_df$Term <- factor(smp_go_top10_df$Term, levels = rev(smp_go_top10_df$Term))

data <- smp_go_table[1:11,]


data <- data %>%
        mutate(
        Combined_score = data$`Combined Score`,
        EnrichmentFDR = -log10(data$`Adjusted P-value`),
        N_genes = sapply(strsplit(as.character(Genes), ";"), length)
        )

data <- data[data$EnrichmentFDR > 1.30103 ,]
plot <- ggplot(data, aes(x = reorder(Term, Combined_score), y = Combined_score)) +
        geom_point(aes(size = N_genes, color = EnrichmentFDR)) +
        scale_color_gradient(low = "blue", high = "red") +
        coord_flip() +
        labs(
        title = "Top10 GO term",
        x = "",
        y = "Combined score",
        color = "EnrichmentFDR",
        size = "Genes"
        ) +
        theme_minimal() +
        theme(axis.text.x = element_text(size = 15),
              axis.text.y = element_text(size = 15),
              axis.text = element_text(size = 15,colour = "black"),
              axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
              axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
              axis.title = element_text(size = 15,colour = "black"),
              legend.title = element_text(size = 15),
              legend.text = element_text(size = 15))
  
  plot
# ggsave(plot = plot,filename = paste0(fig_proj,"fig4A_filter_dmr_smp_GO_top10_p_20240722.pdf"),
#          device = "pdf",width = 15, height = 6)
  
```

## fig4b  Add DMRS region for the annotated genes .Heatmap of  intrested GO 4  gene terms(20240829 modified)
```{r,fig.width=15,fig.height=10}
## import data
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/code/gene_anlyais/"
go_gene_4 <- read.csv(paste0(dat_dir,"Go4_genes_meth.csv"))
## 
go4_gene_p <- go_gene_4[go_gene_4$p_adjust<0.05,]

##sort datafram arrange genes
order_name <- c( "DPYSL3","RAPGEF2", "NLGN1", #duplicate gene
                  "TENM3" ,"MAGI2","ROBO1","SLIT2","EPHA3","DDR2","NRXN1","CDH12",  
                  "SLITRK6","FN1","PKP3","CDH9","POU4F1","CDH8","ADGRL3","SMARCB1","ARID1B","SMARCA2","KANK1", 
                  "DCC","LPAR1" )

go_gene_4$Genes <- factor(go_gene_4$Genes, levels = order_name)

sorted_go_gene_4 <- go_gene_4 %>% arrange(Genes)
##add new cols contain gene_chr_start_end information 

sorted_go_gene_4[,4:8] <- lapply(sorted_go_gene_4[,4:8],as.character)
sorted_go_gene_4$new_col <- apply(sorted_go_gene_4[,c(2,5,4,6)],1,function(x) paste(x,collapse = "_"))

## extract the data do heatmap
heatmap_dat <- sorted_go_gene_4[,10:48]
rownames(heatmap_dat) <- heatmap_dat$new_col
gp_mat <- heatmap_dat[,1:17]%>%as.matrix()
rownames(gp_mat) <- rownames(heatmap_dat)
gc_mat <- heatmap_dat[,18:38]%>%as.matrix()
rownames(gc_mat) <- rownames(heatmap_dat)
## annotation setting 
type_colors <- c("GC" = "#FB8D8D", "GP" = "#C0E1FF")
status_colors <- c("Control" = "#63A6D0", "Early Stage" = "#F9A68A", "Advanced Stage" = "#D34447")
gender_colors <- c("Male" = "#AACBFE", "Female" = "#DE8686")
my_color=c("#4675FE","#F9F5C9","#FCA605")
row_groups <- c(rep("GO:0031346;GO:0031345", 2),
                rep("GO:0031346;GO:0034329",1),
                rep("GO:0031346",6),
                rep("GO:0034329",9),
                rep("GO:0045582",3),
               rep("GO:0031345",3))
## select annotated DMR region CHD12 enhancer  
row_enhancer <- c(rep("No",10),"Yes",rep("No",12),"Yes")%>%as.factor()
# row_ann <- rowAnnotation(GO=row_groups,col =list(GO=row_group_col),labels=GO_labels)
ann_ht1 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[1:17],
                            Gender=meta_df$sc_gender[1:17],
                            Type=meta_df$sc_type[1:17],
                            annotation_name_side = "left",
                            col=list(
                                  Type = type_colors,
                                  Status = status_colors,
                                  Gender = gender_colors))

ann_ht2 <- HeatmapAnnotation(
                            Status=meta_df$sc_stage[18:38],
                            Gender=meta_df$sc_gender[18:38],
                            Type=meta_df$sc_type[18:38],
                            col=list(Type = type_colors,
                                      Status = status_colors,
                                      Gender = gender_colors),
                                      annotation_name_side = "right")
row_enhacer_col <- rowAnnotation(Enhancer=row_enhancer,col=list(Enhancer=c("No"="#999999","Yes"="#F2F2F2")))
row_annots <- rowAnnotation(GO = row_groups,
                           col = list(GO = c("GO:0034329"="#2F7EBC",

                                             "GO:0031345"="#F96949",

                                             "GO:0031346"="#50B8DD",

                                              "GO:0045582"="#FCBC85",

                                             "GO:0031346;GO:0034329"="#77C8A6",

                                             "GO:0031346;GO:0031345"="#FCEBDE"
                                            )),
                            annotation_legend_param = list(
                            GO = list(
                            at = c("GO:0034329", "GO:0031345", "GO:0031346", "GO:0045582", 
                                    "GO:0031346;GO:0034329", "GO:0031346;GO:0031345"),
                            labels = c("Cell Junction Assembly", 
                                          "Positive Regulation Of T Cell Differentiation", 
                                          "Positive Regulation Of Cell Projection Organization", 
                                          "Negative Regulation Of Cell Projection Organization", 
                                          "Positive Regulation Of Cell Projection Organization;Cell Junction Assembly", 
                                          "Positive Regulation Of Cell Projection Organization;Negative Regulation Of Cell Projection Organization")
                             )
                           ))
## do heatmap
# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_top10_gene_heatmap_20240731.pdf"),width = 16,height = 10)
heat_map_1 <- Heatmap(gp_mat,name="gp",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                     # right_annotation = row_annots,
                      top_annotation = ann_ht1,
                      show_column_names = F,
                     show_row_dend = F,
                     show_row_names = T,
                     row_title = NULL,
                       col = colorRampPalette(my_color)(100),
                       na_col = "white",width =unit(4,"cm"),
                     height = unit(10,"cm")
                      )
heat_map_2 <- Heatmap(gc_mat,name="gc",
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                      right_annotation =c(row_annots,row_enhacer_col) ,
                      top_annotation = ann_ht2,
                      show_column_names = F,
                      show_row_dend = F,
                      show_row_names = T,
                       row_title = NULL,
                      col = colorRampPalette(my_color)(100),
                      na_col = "white",width = unit(4,"cm"),
                      height = unit(10,"cm"),
                      row_dend_side = "right")


p <- heat_map_1+heat_map_2

# pdf(paste0(fig_proj,"fig4B_filter_dmr_smp_GO_spe4_gene_heatmap_20240829.pdf"),width = 16,height = 10)
draw(p, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")


# dev.off()

```
##fig 4c equal to fig3g-h

##fig 4e GO top 10 term barplot 20240722 modified 
```{r,fig.width=20,fig.height=8}
work_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/figS5-6_dat/"
scstage_go_table <- read_excel(paste0(work_dir,"GO_Biological_Process_2023_table (2).xlsx"),sheet = "GO_Biological_Process_2023_tabl")
# scstage_go_df <- scstage_go_table[1:10,c(1,3,4)]
# scstage_go_df$log10p_adjust <- -log(scstage_go_df$`Adjusted P-value`)
data <- scstage_go_table[1:10,]

data <- data %>%
    mutate(
      Combined_score = data$`Combined Score`,
      EnrichmentFDR = -log10(data$`Adjusted P-value`),
      N_genes = sapply(strsplit(as.character(Genes), ";"), length)
      )
 p4 <- ggplot(data, aes(x = reorder(Term, Combined_score), y = Combined_score)) +
    geom_point(aes(size = N_genes, color = EnrichmentFDR)) +
    scale_color_gradient(low = "blue", high = "red") +
    coord_flip() +
    labs(
      title = "Top10 GO term",
      x = "",
      y = "Combined score",
      color = "EnrichmentFDR",
      size = "Genes"
    ) +
    theme_minimal() +
   theme(axis.text.x = element_text(size = 15),
                    axis.text.y = element_text(size = 15),
                    axis.text = element_text(size = 15,colour = "black"),
                    axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.title = element_text(size = 15,colour = "black"),
                    legend.title = element_text(size = 15),
                    legend.text = element_text(size = 15))
  
# ggsave(plot = p4,filename = paste0(fig_dir,"figS5D_scstage_GO_top10_terms_20240722.pdf"),
#          device = "pdf",width = 15, height = 6)
print(p4)
```

## fig4 f-g all scstage_dmr associated gene violin plot(20240829 modified)
```{r,echo=TRUE}
##----------------------------------------DO violin plot for the associated DMRS gene ---------------------------------------
## import data
gene_dat <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/result/scstage_change_ann/"
# scstage_meth_dmr <- read.csv(paste0(gene_dat,"scstage_meth_data.csv"))
# scstage_meth_dmr$p_adjust <- as.numeric(scstage_meth_dmr$p_adjust)

smp_scsatge_list <- c("sc01_14-Z","sc02_L14-1502","sc04_L14-24",
                      "sc05_L15-24","sc06_L15-33","sc07_L15-25","sc08_L15-1608","sc09_L19-1","sc10_L19-2","sc11_L19-3",
                      "sc12_L19-4","sc13_L11-2","sc14_L11-5","sc15_L12-2","sc16_L12-3","sc17_L12-4","sc18_L12-5",
                       "sc19_L12-6","sc20_L12-8","sc21_L16-1","sc23_L16-6")

smp_scstage_DMR_meth <- read.table(paste0(sc_dat_dir,"scstage_dmr_change_region.sorted.txt"),col.names = "chr_region")
smp_scstage_DMR_cov <- read.table(paste0(sc_dat_dir,"scstage_dmr_change_region.sorted.txt"),col.names = "chr_region")

for (i_sample in smp_scsatge_list){
  a <- read.table(paste0(sc_dat_dir,i_sample,"_sc_dmr.calc.bed"),col.name = c("chromosome","start","end","mod","unmod","meth_level"))

 smp_scstage_DMR_meth[[i_sample]] <- a$meth_level
 smp_scstage_DMR_cov[[i_sample]] <- a$mod+a$unmod
}
## scstage_meth_dmr 2414 dmr_region
scstage_meth_dmr <- read.csv(paste0(gene_dat,"scstage_meth_data.csv"))
scstage_meth_dmr$p_adjust <- as.numeric(scstage_meth_dmr$p_adjust)
scstage_dat_ann <- scstage_meth_dmr

## trans data format
scstage_dat_ann[,c(1:5)] <- lapply(scstage_dat_ann[,c(1:5)], as.character)
scstage_dat_ann$chr_region <- apply(scstage_dat_ann[,c(1:3)], 1, function(x) paste(x,collapse = "_"))
scstage_dat_ann$gene_region <-  apply(scstage_dat_ann[, c(1:5)], 1, function(x) paste(x, collapse = "_"))
## define scstage_dmr gene annotated
scstage_dmr_ann <- scstage_dat_ann[,c(27,28,6:26)]

## merge data
merge_scstage_meth_ann <- merge(scstage_dmr_ann,smp_scstage_DMR_meth,by="chr_region")
scstage_meth_ann <- merge_scstage_meth_ann[,c(2,24:44)]

merge_scstage_cov_ann <- merge(scstage_dmr_ann,smp_scstage_DMR_cov,by="chr_region")
scstage_cov_ann <- merge_scstage_cov_ann[,c(2,24:44)]

## scstage_meth
data <- scstage_meth_ann
rownames(data) <- as.character(data$gene_region)
data <- data[, -1]
tr_data <- t(data)
rownames(tr_data) <- colnames(data)
colnames(tr_data) <- rownames(data)
tr_data_meth <- as.data.frame(tr_data)
# tr_data_meth$smp_type <- meta_df$sc_type

## scstage_cov
data_cov <- scstage_cov_ann
rownames(data_cov) <- as.character(data_cov$gene_region)
data_cov <- data_cov[, -1]
tr_data_co <- t(data_cov)
rownames(tr_data_co) <- colnames(data_cov)
colnames(tr_data_co) <- rownames(data_cov)
tr_data_cov <- as.data.frame(tr_data_co)
##

common_cols <- intersect(names(tr_data_cov), names(tr_data_meth))
count=0
p_adjust <- list()
p_value <- list()
## 
for (col in common_cols) {
  df <- data.frame(x = tr_data_cov[[col]], y = tr_data_meth[[col]]) 
  df$status <- meta_scstage_df$stage
  df$smp <- str_sub(rownames(tr_data_cov),1,4)
  df <- na.omit(df)
  ## do scatter plot 
    wilcox_test <- wilcox.test(df$y[df$status=="Early Stage"],df$y[df$status=="Advanced Stage"],paired = FALSE)
    p_value <- wilcox_test$p.value
    p_adjust[[col]] <- p_value
  if(!is.na(p_value)&&p_value< 0.05&&length(p_value)!=0){ 
        p1 <- ggplot(df, aes(x = status, y = y)) +
              geom_violin(aes(fill=status,alpha=1),scale="width") +
              scale_fill_manual(values =c("#D34447", "#F9A68A"))+
              geom_boxplot(aes(fill=status),
                            outlier.shape = NA,lwd=0.4,width=0.05)+
              scale_fill_manual(values =c("#FB4F4F", "#FE7E2A"),guide = 'none')+
              labs(y ="Methylation level", x = paste("y_", col), title = paste0("p = ", round(p_value, 10)))+
              theme_classic()+
              theme(axis.text.x = element_text(size = 15),
                    axis.text.y = element_text(size = 15),
                    axis.text = element_text(size = 15,colour = "black"),
                    axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
                    axis.title = element_text(size = 15,colour = "black"),
                    legend.position="none")
        # ggsave(plot = p1,
        #           filename =paste0(fig_dir,"scstage_dmr_gene_violin_plot_20240902/meth_",col,"_violin_plot.pdf"),
        #        width = 6,height = 4,dpi = 300)
        print(p1)
         }else{
             print(paste0(col,"_wilcox p.adjust less than 0.05,and the p is ",p_value))
            count=count+1 }
          print(count)
 }




```

## technical
```{r}
sessionInfo()
```

