---
title: "fig2_technical_analysis"
author: "Zhenna Jiao"
date: "2024-03-04(last modified: `r Sys.Date()`)"
output: 
    prettydoc::html_pretty:
    theme: cayman
    highlight: github
---
# load packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(readxl)
library(ggpubr)
library(stringr)
library(readr)
library(Rmisc)
library(dplyr)
library(smoother)
library(reshape2)
library(matrixStats)
library(RColorBrewer)
library(ComplexHeatmap)
```
## Data and result stored
```{r}
fig_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/technical/figure/fig2/"
proj <- "/home/jiaozhenna/methylation/project/EM_seq/data/metadata/"
technical_dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/technical/data/"
```
## figS4 a-b The violin plot for samples input concentration(EV-DNA in patients plasma) and coverage 
```{r}
##import data
input_df <- read_excel(paste0(technical_dat_dir,"downsample.xlsx"),sheet = "downsample")

dat <- input_df[,c(1,19:20)]
rownames(dat) <- dat$sample_name...1
colnames(dat) <- c("smp","plasma_conc","coverage")
state <- dat$smp
dat$status <- substr(state,1,2)%>%as.factor()
dat$type <- c(rep("GP",21),rep("GC",23))%>%as.factor()
p_value_cov <- wilcox.test(dat$coverage[1:21],dat$coverage[22:44])[["p.value"]]

## do plot for coverage difference between gp and gc 
p1 <- ggplot(dat, aes(x =type, y =coverage,col=type)) +
          geom_violin(aes(fill=type,alpha=0.5),scale="width")+
          scale_fill_manual(values=c("#FB8D8D","#C0E1FF"))+
          geom_boxplot(outlier.shape = NA,lwd=0.8,width=0.05)+
          geom_jitter(position=position_jitter(0.1),aes(fill=type))+
          scale_color_manual(values=c("#FB8D8D","#C0E1FF"))+
          theme_classic()+
          ylim(0,1)+
          labs(x ="", y = "Coverage Rate", 
                title = paste0("coverage rate between gp and gc ,the p.value is ",round(p_value_cov,3)))+
          theme(axis.text.x = element_text(size = 15,colour = "black"),
                axis.text.y = element_text(size = 15,colour = "black"),
                axis.title = element_text(size = 15,colour = "black"),
                axis.line = element_line(size=1,linetype = 1,color = "black"))
        
      
# ggsave(plot = p7,filename =paste0(fig_dir,"figS2F_1_downsample_sample_cov_20240902.pdf"),
#        width = 8,height = 6 ,dpi = 300)
print(p1)

## do plot for input concentration difference between gp and gc 
p_value_input <- wilcox.test(dat$plasma_conc[1:21],dat$plasma_conc[22:44])[["p.value"]]

p2 <- ggplot(dat, aes(x =type, y =plasma_conc,col=type)) +
          geom_violin(aes(fill=type,alpha=0.5),scale="width")+
          scale_fill_manual(values=c("#FB8D8D","#C0E1FF"))+
          geom_boxplot(outlier.shape = NA,lwd=0.8,width=0.05)+
          geom_jitter(position=position_jitter(0.1), aes(fill=type))+
          scale_color_manual(values=c("#FB8D8D","#C0E1FF"))+
          theme_classic()+
          labs(x =" ", y = "Input Concentration (ng/ml)", 
             title = paste0("input concentration between gp and gc ,
                            the p.value is ",round(p_value_input,3)))+
          theme(axis.text.x = element_text(size = 15,colour = "black"),
                axis.text.y = element_text(size = 15,colour = "black"),
                axis.title = element_text(size = 15,colour = "black"),
                axis.line = element_line(size=1,linetype = 1,color = "black"))
       
# ggsave(plot = p8,filename =paste0(fig_dir,"figS2F_downsample_sample_plasma_concentration_20240902.pdf"),
# width = 8,height = 6 ,dpi = 300)
print(p2)

```
## figS5 a The barplot for samples age distribution between control early stage and advanced stage 
```{r,fig.width=8,fig.height=6}
## import data
age_gender_df <- read_excel(paste0(technical_dat_dir,"age_and_gender.xlsx"),sheet = "filter_sample")
age_gender_df$Stage <- factor(age_gender_df$Stage,levels = c("Control","Early Stage","Advanced Stage"))
## do boxplot
p3 <-ggplot(age_gender_df, aes(x=Stage,y=Age))+
    geom_boxplot(outlier.shape = NA,lwd=0.6,
                 aes(fill=Stage),alpha=0.9)+
    geom_jitter(position=position_jitter(0.1), 
                aes(colour=Gender))+
    theme_classic()+
    labs(x = "", y = "Age",lwd = 1.5)+
    ylim(0,100)+
    scale_color_manual(values=c("#DE8686","#AACBFE"))+
    scale_fill_manual(values = c("#63A6D0","#F9A68A","#D34447"))+
    theme(axis.text.x = element_text(size = 15,colour = "black"),
          axis.text.y = element_text(size = 15,colour = "black"),
          axis.title = element_text(size = 15,colour = "black"),
          axis.line = element_line(size=1,linetype = 1,color = "black"),
          legend.text = element_text(size = 15), 
          legend.title = element_text(size = 15), 
          plot.title = element_text(size = 15),
          legend.position = "right")
# ggsave(plot = p6,filename = paste0(fig_dir,"figS2E_boxplot_age_gender.pdf"),width = 8,height = 6,dpi = 300)
print(p3)
```
## figS5 b-c the distribution of %CpG modified level in 1million region in different chromosomes
```{r}
##import data
merge_df <- readRDS("/home/jiaozhenna/methylation/project/EM_seq/analysis/technical/data/ds_1kb_count.rds")
## define the function to do plot for the methylation level in million region on different chromosomes 
chr_variance <- function(data,chromosome){
mat2 <- data
mat<-mat2%>%dplyr::filter(chr==chromosome)
#prepare 1 mb windows
ws<-1000
mat$bin<-c(1:nrow(mat))
mat<-data.frame(mat)
#index widows
for (i in grep ("bin", colnames(mat))){
  mat$index<-ceiling(mat[,i]/ws)
}
mat[is.na(mat)]<-0
mat<-sapply(mat[,-1], as.numeric)
mat<-data.frame(mat)
mat.100kb<-aggregate(mat[, grep("uC|mC|nC", colnames(mat))],
                     by = list(mat$index),
                     FUN = sum)
for (indexmC in grep("_mC", colnames(mat.100kb))) { #index columns containing "_mC" in colname
  indexuC <- grep(gsub("_mC", "_uC", colnames(mat.100kb)[indexmC]), colnames(mat.100kb)) #index columns with uC
  m<-mat.100kb[,indexmC]/(mat.100kb[,indexuC]+mat.100kb[,indexmC])
  n<-ncol(mat.100kb)
  mat.100kb<-data.frame(mat.100kb,m)
  colnames(mat.100kb)[n+1]<-gsub("_mC", "_mod", colnames(mat.100kb)[indexmC]) 
}  

mod_df=mat.100kb[,grep("_mod",colnames(mat.100kb))]
mod_df

##smooth
smooth<-data.frame(mat.100kb$Group.1)
for (i in 1:ncol(mod_df)){
  name<-colnames(mod_df)[i]
  tmp<-mod_df[,i]
  tmp.smooth<-smth.gaussian(tmp,
                            window = 10,
                            tails=TRUE,
                            alpha=1)
  n<-ncol(smooth)
  smooth<-data.frame(smooth,tmp.smooth)
  colnames(smooth)[n+1]<-name
}
healthy<-smooth[,grep("gp", colnames(smooth))]
healthy$bin<-c(1:nrow(healthy))
GC <-smooth[,grep("sc", colnames(smooth))]
GC$bin<-c(1:nrow(GC))
healthy.melt<-melt(healthy, id="bin")
GC.melt<-melt(GC,id="bin")
colours=colorspace::sequential_hcl(30)

p3<-ggplot(healthy.melt, aes(x=bin, y=value, colour=variable))+
  geom_line()+
  theme_classic()+
  ggtitle("GP") +
  xlab(paste0("Postion on ",chromosome,"(x10^6)")) +
  ylab("Methylation value (%)")+
  ylim(0.5,0.9)+
  xlim(0,200)+
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5, vjust = 0.5),
        axis.text.x = element_text(size = 15,  hjust = 0.5,color = "black"),
        axis.text.y = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_text(color = "black", size = 15),
        axis.title.y = element_text(color = "black", size = 15),
        axis.line = element_line(size=1,linetype = 1,color = "black"))
# ggsave(plot = p3,filename = paste0(fig_dir,"figS2C_GP_chr4_meth_dist.pdf"),
#        width = 6,height = 4,dpi=300)
print(p3)

p4<-ggplot(GC.melt, aes(x=bin, y=value, colour=variable))+
  geom_line()+
  theme_classic()+
  ggtitle("GC") +
  xlab(paste0("Postion on ",chromosome," (x10^6)")) +
  ylab("Methylation value (%)")+
  ylim(0.5,0.9)+
  xlim(0,200)+
  theme(legend.position = "none",
         plot.title = element_text(hjust = 0.5, vjust = 0.5),
        axis.text.x = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.text.y = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_text(color = "black", size = 15),
        axis.title.y = element_text(color = "black", size = 15),
        axis.line = element_line(size=1,linetype = 1,color = "black"))
# ggsave(plot = p4,filename = paste0(fig_dir,"figS2C2_GC_chr4_meth_dist.pdf"),
#        width = 6,height = 4,dpi = 300)
print(p4)
}

## plot the methylation level on chromosome 4
i=4
chr_variance(data = merge_df ,chromosome = paste0("chr",i))

## figS2B methylation level variance  in 1mb genome between GP and GC(the code ref to taps)
mat2<- merge_df
meth<- merge_df
meth$idx <- floor(meth$start / 1000000)  # generate index for each 1,000,000 bp bin
meth[is.na(meth)]<-0
meth<-sapply(meth[,-1], as.numeric)
meth<-data.frame(meth)
meth$Chr<-mat2$chr

counts <- aggregate(meth[, grep("mC|nC|uC", colnames(meth))], by = list(meth$Chr, meth$idx), FUN=sum)  

#calculate methylation
for (indexmC in grep("_mC", colnames(counts))) { #index columns containing "_mC" in colname
  indexuC <- grep(gsub("_mC", "_uC", colnames(counts)[indexmC]), colnames(counts)) #index columns with uC
  m<-counts[,indexmC]/(counts[,indexuC]+counts[,indexmC])
  n<-ncol(counts)
  counts<-data.frame(counts,m)
  colnames(counts)[n+1]<-gsub("_mC", "_mod", colnames(counts)[indexmC]) 
}  

autosomes<-c("chr1$","chr2$","chr3$","chr4$","chr5$","chr6$","chr7$","chr8$","chr9$","chr10$",
             "chr11$","chr12$","chr13$","chr14$","chr15$","chr16$","chr17$","chr18$","chr19$","chr20$","chr21$","chr22$")

#select autosomes
counts<-counts[grep(paste(autosomes, collapse = "|"),counts$Group.1),]


mod_df=counts[,grep("_mod",colnames(counts))]
mod_df$pos<-paste(counts$Group.1, counts$Group.2, sep="_")
mod_df<-mod_df[complete.cases(mod_df),]
colnames(mod_df)<-gsub("_mod","",colnames(mod_df))

GP <-mod_df[,grep("gp", colnames(mod_df))]
rownames(GP)<-mod_df$pos
GC <-mod_df[,grep("sc", colnames(mod_df))]
rownames(GC)<-mod_df$pos
GP.var <-rowVars(as.matrix(GP))
GC.var<-rowVars(as.matrix(GC))
var.df<-data.frame(mod_df$pos,GP.var,GC.var)
var.df<-var.df%>%tidyr::separate(mod_df.pos,c("chr","pos"),"_")
var.melt<-melt(var.df,id.vars = c("chr","pos"))
chrOrder <-c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11",
             "chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20",
             "chr21","chr22")
var.melt$chr <- factor(var.melt$chr, chrOrder, ordered=TRUE) 
var.melt$status <- substr(var.melt$variable,1,2)
var.melt$status <- factor(var.melt$status,level=c("GP","GC"))

p5 <- ggplot(var.melt, aes(x =status, y =value)) +
        geom_boxplot(aes(fill=status),outlier.shape = NA,
                     lwd=0.7,alpha=0.8,width=0.5)+
        theme_classic()+
        ylim(0,0.001)+
        scale_fill_manual(values=c("#FB8D8D","#C0E1FF"))+
        labs(x = "", y = "Methylation variance",lwd = 1.5)+
        theme(
          axis.text.x = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.text.y = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.title.x = element_text(color = "black", size = 15),
        axis.title.y = element_text(color = "black", size = 15),
        axis.line = element_line(size=1,linetype = 1,color = "black"),
          legend.text = element_text(size = 15), 
          legend.title = element_text(size = 15), 
          plot.title = element_text(size = 15),
          legend.position = "none"
          )

# ggsave(p5,
#        filename=paste0(fig_dir,"figS2B_smp_mod_variance.pdf"),
#        width=6,
#        height = 4,
#        dpi = 300)
print(p5)
```

## figS5 d the distribution of %CpG modified level 
```{r}
##import data
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/result/technical/"
sample_names <- c("gp01_L14-1550" ,"gp03_L15-31","gp04_L15-12","gp05_L15-1550",
                   "gp06_L15-13","gp07_L19-5","gp08_L19-6","gp09_L19-7","gp10_L19-8","gp11_L11-1",
                   "gp12_L11-3","gp14_L11-8","gp15_L12-1","gp17_L16-2","gp18_L16-4",
                   "gp19_L16-7","gp20_L16-8","sc01_14-Z","sc02_L14-1502","sc04_L14-24",
                   "sc05_L15-24","sc06_L15-33","sc07_L15-25","sc08_L15-1608","sc09_L19-1","sc10_L19-2","sc11_L19-3",
                   "sc12_L19-4","sc13_L11-2","sc14_L11-5","sc15_L12-2","sc16_L12-3","sc17_L12-4",
                  "sc18_L12-5","sc19_L12-6",
                   "sc20_L12-8","sc21_L16-1","sc23_L16-6")

smp_chr_data <- readRDS(paste0(dat_dir,"smp_chr_data.rds"))
mod_count <- readRDS(paste0(dat_dir,"mod_count.rds"))

df <- do.call(rbind, lapply(names(mod_count), function(name) mod_count[[name]]))
df$mod <- df$mod/100

df$bin<-NA
df$bin<-ifelse(df$mod>=0 & df$mod<=0.1,1,
               ifelse(df$mod>0.1 & df$mod<=0.2,2,
                      ifelse(df$mod>0.2 & df$mod<=0.3,3,
                             ifelse(df$mod>0.3 & df$mod<=0.4,4,
                                    ifelse(df$mod>0.4 & df$mod<=0.5,5,
                                           ifelse(df$mod>0.5 & df$mod<=0.6,6,
                                                  ifelse(df$mod>0.6 & df$mod<=0.7,7,
                                                         ifelse(df$mod>0.7 & df$mod<=0.8,8,
                                                                ifelse(df$mod>0.8 & df$mod<=0.9,9,
                                                                       ifelse(df$mod>0.9 & df$mod<=1,10,NA))))))))))

df$status<-NA
df$status[grep("gp",df$name)]<-"GP"
df$status[grep("sc",df$name)]<-"GC"
# df<-df[-grep("GP|GC", df$name),]
df$name<-as.character(df$name)
df<-aggregate(count ~ bin + name, FUN=sum, data=df)
sums<-aggregate(df$count, by=list(df$name), FUN=sum)
colnames(sums)<-c("name","total")
df <- merge(x = df, y = sums, by = "name")
df$freq<-(df$count/df$total)*100


df$status<-NA
df$status[grep("gp",df$name)]<-"GP"
df$status[grep("sc",df$name)]<-"GC"
# df$status[grep("Healthy_",df4$name)]<-"Healthy"

df2<-aggregate(freq ~ bin + status, FUN=mean, data=df)
df2$status<-factor(df2$status, levels=c("GP","GC"), ordered=TRUE) 

df$mean_freq<-NA
df3 <- df %>% right_join(df2, by=c("bin","status"))
df3$status<-factor(df3$status, levels=c("GP","GC"), ordered=TRUE)

## do hist plot 
b.mod_gp <- ggplot(df3[df3$status == "GP",]) +
  geom_bar(aes(x=bin, y=freq.y), position="identity", stat="identity", color="black", fill="#C0E1FF") +
  geom_line(aes(x=bin, y=freq.x, colour=name), position="identity", stat="identity") +
  theme_classic() +
  ggtitle("GP")+
  ylab("% of modified CpG sites") +
  xlab("Modification level (%)")+
  # facet_wrap(~ status,scales = "free_x") +
  theme(legend.position = "none", 
        strip.text = element_text(face = "bold", size = 15, color = "black", angle = 0),
        strip.background = element_blank(), 
        axis.text.x = element_text(size = 15, angle = 90, hjust = 0.5,color = "black"),
        axis.text.y = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_text(color = "black", size = 15),
        axis.title.y = element_text(color = "black", size = 15),
        axis.line = element_line(size=1,linetype = 1,color = "black"),
        plot.title = element_text(hjust = 0.5)) +
        scale_x_continuous(breaks = seq(1, 10, by = 1),
                      labels = c( "0-10", "10-20", "20-30", "30-40", 
                                  "40-50", "50-60", "60-70", "70-80", "80-90", "90-100"))

# ggsave(paste0(fig_dir,"figS2D_plot_GP_mod_level.pdf"),plot = b.mod_gp,width = 4,height =4,dpi = 300)
print(b.mod_gp)

b.mod_gc <- ggplot(df3[df3$status == "GC",]) +
  geom_bar(aes(x=bin, y=freq.y), 
           position="identity", stat="identity", color="black", fill="#FB8D8D") +
  geom_line(aes(x=bin, y=freq.x, colour=name), 
            position="identity", stat="identity") +
  theme_classic() +
  ylab("% of modified CpG sites") +
  xlab("Modification level (%)")+
  ggtitle("GC")+
  # facet_wrap(~ status,scales = "free_x") +
  theme(legend.position = "none", strip.text = element_text(face = "bold", size = 15, color = "black", angle = 0),
        strip.background = element_blank(), 
        axis.text.x = element_text(size = 15, angle = 90, hjust = 0.5,color = "black"),
        axis.text.y = element_text(size = 15,hjust = 0.5,color = "black"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_text(color = "black", size = 15),
        axis.title.y = element_text(color = "black", size = 15),
        axis.line = element_line(size=1,linetype = 1,color = "black"),
        plot.title = element_text(hjust = 0.5)) +
        scale_x_continuous(breaks = seq(1, 10, by = 1),
                        labels = c( "0-10", "10-20", "20-30", "30-40",
                                    "40-50", "50-60", "60-70", "70-80", "80-90", "90-100"))


# ggsave(paste0(fig_dir,"figS2D_plot_GC_mod_level.pdf"),plot = b.mod_gc,width = 4,height =4,dpi = 300)
print(b.mod_gc)
```
## figS6 a do correlation heatmap about DMR which called between Early and Advanced Stage 
```{r,fig.width=14,fig.height=10}
## set fig_stored dir
fig_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/figures/figS5-6/"
meta_scstage_df <- readRDS("/home/jiaozhenna/methylation/project/EM_seq/data/metadata/meta_scstage_df.rds")
sc_dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/filter_sample/scstage_dmr_change/"
## import sc_satge data
scstage_DMR_meth <- readRDS(paste0(sc_dat_dir,"scstage_DMR_change_meth.rds"))
## calculate pearson correlation coefficient matrix
data_keep_na <-  scstage_DMR_meth
row.names(data_keep_na) <- as.character(scstage_DMR_meth$chr_region)
data_keep_na <- data_keep_na[,-1]
colnames(data_keep_na) <- meta_scstage_df$smp
DMR_cor_keep_na <- cor(data_keep_na,use = "pairwise.complete.obs")
annotation_col_keepna <- data.frame(Stage=meta_scstage_df$stage,Gender=meta_scstage_df$gender)
DMR_cor_df_keepna <- as.data.frame(DMR_cor_keep_na)
rownames(annotation_col_keepna)  <- colnames(DMR_cor_df_keepna)

my_color=c("#4675FE","white","#F9A68A","#D34447")

gender_color=list(Gender=c(Male="#AACBFE",Female="#DE8686"))
stage_color=list(Stage=c("Early Stage"="#F9A68A","Advanced Stage"="#D34447"))

# pdf(paste0(fig_dir,"figS5A_heatmap_scstage_dmr_cor.pdf"),width = 14,height = 10)
pheatmap::pheatmap(DMR_cor_df_keepna,cluster_rows = T ,cluster_cols =T,show_rownames = T,show_colnames = T,
                   annotation_col = annotation_col_keepna,
                   annotation_colors = c(gender_color,stage_color),
                   scale = "none",
                  fontsize = 15,                
                  fontsize_row = 15,             
                  fontsize_col = 15,
                   cellwidth = 13,
                  cellheight =12,
                  color = colorRampPalette(my_color)(100))

# dev.off()
```

## figS6 b do pca about DMR which called between Early and Advanced Stage(the data use dmr_meth remove NA)

```{r}
## import  data remove missing value
dat_rm_na <- scstage_DMR_meth%>%na.omit()
## trans data format
rownames(dat_rm_na) <- as.character(dat_rm_na$chr_region)
dat_rm_na <- dat_rm_na[, -1]
tr_data <- t(dat_rm_na)
rownames(tr_data) <- colnames(dat_rm_na)
colnames(tr_data) <- rownames(dat_rm_na)
tr_data <- as.data.frame(tr_data)
tr_data$sample_stage <- meta_scstage_df$stage %>% as.factor()

## calculate PCs
tr.pca_all <- prcomp(tr_data[,1:(length(tr_data)-1)])

## do plot 
df1 <- tr.pca_all$x
pc_one <- as.data.frame(df1,tr_data$sample_stage)
pc_one$Stage <- tr_data$sample_stage
pc_one$smp_name <- substr(rownames(tr_data),1,4)
#Extract the variance contribution rate of the principal components and generate axis titles；
summ<-summary(tr.pca_all)
xlab<-paste0("PC1(",round(summ$importance[2,1]*100,2),"%)")
ylab<-paste0("PC2(",round(summ$importance[2,2]*100,2),"%)")
p1 <-ggplot(data = pc_one,aes(x=PC1,y=PC2,color=Stage))+
  geom_point(size=4.5)+
  labs(x=xlab,y=ylab)+
  theme_classic()+
  scale_color_manual(values = c("#D34447", "#F9A68A"))+
  theme(axis.text.x = element_text(size = 15),
        axis.text.y = element_text(size = 15),
        axis.text = element_text(size = 15,colour = "black"),
        axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
        axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
        axis.title = element_text(size = 15,colour = "black"))
 

# ggsave(paste0(fig_dir,"figS6_PCA_scstage_dmr_removena.pdf"),plot = p1,width = 8,height = 5,dpi = 300)
print(p1)
```
## figS6 c fisher result dat from shell which sorted in scstage dmr region
```{r}
proj_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/result/annotation_dmr/"
dat_sc <- read_excel(paste0(proj_dir,"dmr_features.xlsx"),sheet = "scstage_dmr_feature_fisher")
dat_sc$logp <- -log(dat_sc$two_tail,10)

p4 <- ggplot(dat_sc, aes(y= elments, x= logp)) +
      geom_bar(stat = "identity",fill="#F9A68A",color = "#F9A68A", width = 0.8) +
      theme_classic()+
      ylab("-log10(Pvalue)") +
      xlab("")+
      theme(axis.text.x = element_text(size = 15),
            axis.text.y = element_text(size = 15),
            axis.text = element_text(size = 15,colour = "black"),
            axis.line.x = element_line(size = 1,linetype = 1,colour = "black"),
            axis.line.y = element_line(size = 1,linetype = 1,colour = "black"),
            axis.title.x = element_text(size = 15,colour = "black"))
# ggsave(paste0(fig_dir,"lolipop_scstage_dmr_20240722.pdf"),plot = p4,width = 7,height = 4,dpi = 300)
print(p4)
```

## figS7 a-b
```{r}
##--------------Randomforest prediction-------------------------------------------
## the code equal to fig3 e
```

## figS8
```{r}
## extract top  p< 0.05  50 genes
meta_df <- readRDS("/home/jiaozhenna/methylation/project/EM_seq/data/metadata/meta_df.rds")
dat_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/ref_dmr_ann_data/"
meth_dmr <-  read.csv(paste0(dat_dir,"fil_dmr_meth_gene.csv"))
meth_dmr$p_adjust <- as.numeric(meth_dmr$p_adjust)
p_data <- meth_dmr[meth_dmr$p_adjust< 0.05,]

## oreder the dataframe 
p_data <- p_data[order(p_data$p_adjust),]

## extracte top 50 gene rows
h_data <- p_data[1:50,]
## set the data rowname as gene_chr_start_end 
h_data[,c(6,2,3,4)] <- lapply(h_data[,c(6,2,3,4)],as.character)

h_data$new_col <- apply(h_data[,c(6,2,3,4)],1,function(x) paste(x,collapse = "_"))

## extract speciped rows to do heatmap 
top20_gene_dat <- h_data[1:20,7:45]
top20_30_gene_dat <- h_data[21:50,7:45]

## trans datframe as matrix 
rownames(top20_gene_dat) <- top20_gene_dat$new_col
rownames(top20_30_gene_dat) <- top20_30_gene_dat$new_col

## as matrix
top20_gene_dat <- top20_gene_dat[,1:38]%>%as.matrix()
top20_30_gene_dat <- top20_30_gene_dat[,1:38]%>%as.matrix()

## add annotation to the heatmap

annotation_cols <-data.frame(Type=meta_df$sc_type,Stage=meta_df$sc_stage,Gender=meta_df$sc_gender)
rownames(annotation_cols) <- colnames(top20_gene_dat)

# annotation_row <- data.frame(log10p=p_data$p_adjust)

# my_color=c(muted("blue"),"white",muted("red"))
my_color=c("#4675FE","#F9F5C9","#FCA605")

# log10p <- 
# type_colors <- list(Type = c(GP="#4DAF4A", GC= "#FF7F00"))
type_colors=list(Type=c(GP="#C0E1FF",GC="#FB8D8D"))

gender_color=list(Gender=c(Male="#AACBFE",Female="#DE8686"))

stage_color=list(Stage=c("Control"="#63A6D0","Early Stage"="#F9A68A","Advanced Stage"="#D34447"))

# gender_color <- list(Gender=c(Male="#377EB8",Female= "#F781BF"))
# stage_color <- list(Stage=c("Control"="#D6E1F5","Early Stage"="#C2DCAF","Advanced Stage"="#ECDE93"))
annotation_colors <- c(type_colors,gender_color,stage_color) 
##----------------------------heatmap for top_20_30_gene_data----------------------------

pdf(paste0(fig_dir,"fig3F_filter_DMR_heatmap_smp_top20_30_gene_20240829.pdf"),width = 15,height = 6)
pheatmap::pheatmap(top20_30_gene_dat,cluster_rows = T ,
                   cluster_cols =F,show_rownames = T,show_colnames = F,
                   annotation_col = annotation_cols,
                   annotation_colors =annotation_colors,
                   scale = "none",
                    na_col = "white",
                   fontsize =15,
                  fontsize_row = 15,
                  fontsize_col = 15,
                width =unit(38,"cm"),
                height = unit(8,"cm"),
                  # cellwidth = 12,
                  # cellheight =15,
                  color = colorRampPalette(my_color)(100))

dev.off()
```

##figS11 20240802 GO specofied 4 terms associated gene heatmap add element information 
```{r,fig.width=20,fig.height=20}
work_dir <- "/home/jiaozhenna/methylation/project/EM_seq/analysis/DMR_analysis/data/figS5-6_dat/"
scstage_go_element <-read.csv(paste0(work_dir,"scstage_element_go4_meth.csv"))
scstage_speci_go_dup_genes <- read.csv(paste0(work_dir,"scstage_speci4_go_dup_gene_meth.csv"))
scstage_go_element[,c(10,11,12,9)] <- lapply(scstage_go_element[,c(10,11,12,9)], as.character)
scstage_go_element$new_col <- apply(scstage_go_element[, c(10,11,12,9)], 1, function(x) paste(x, collapse = "_"))

h_dat <- scstage_go_element[scstage_go_element$p_adjust<0.05,]

order_name <- c("CDH11","CDH13",##in 4 GO term
                "CDH10", "CDH18", "CDH2", "CDH26","CDH4", "CDH6" ,"CDH7" ,"CDH8", "CDH9",## 3GO term
                "PARD6B","TJP1",#3GO term
               "ADGRL3", "DSCAM", "PCDHB14","SDK2", "SLITRK1",##2GO term
               "CLDN11",
              "CTNND2","PKP2",
               "PRKCA",
               "PTPRT",## GO dup gene
              "TENM2","TENM3","EMB","LRRC4C","GRID2","CADM2",
              "CNTN6","UNC5D","IGSF11","PTPRD","LRFN5","CNTN4","NECTIN1",
              "SLITRK6","WDR1","GJA1","CD9","PTPRU","CLIC4","PTPRJ","MYLK","IGF1R",  
              "DACH1","DUSP10","FGF9","ARID2","PDGFRA","EDN1","ANXA1","HGF","MITF","TIAM1","TMEFF2", 
              "SFRP1","RAP2B","SPATA13","SINHCAF","SGK1","KANK1","LEF1","CXCR4","EGFR","SDCBP", 
              "ERBB4","PDGFD","MGAT5","PDGFC","FYN","CTNNA2","PLXNA4","TCAF2","XBP1","HACE1",  
             "YTHDF3","SULF1","TBX5","CXCL12","STK24","SNAI2","SPRY2","LGR6")

h_dat$Genes <- factor(h_dat$Genes, levels = order_name)
sorted_h_dat <- h_dat %>% arrange(Genes)

## add element info 

sorted_h_dat$elemnets[is.na(sorted_h_dat$elemnets)] <- "None"
sorted_go_dat <- sorted_h_dat[,c(1,9,14:36)]
sorted_h_dat[27,36] <- "chr4_61983982_61984475_ADGRL3_1"
sorted_h_dat[31,36] <- "chr17_73698076_73698175_SDK2_1"
sorted_h_dat[100,36] <- "chr8_58534885_58535347_SDCBP_1"

rownames(sorted_go_dat) <- sorted_h_dat$new_col

heatmap_dat <- sorted_go_dat[,c(3:23)]%>%as.matrix()

status_colors <- c("Control" = "#63A6D0", "Early Stage" = "#F9A68A", "Advanced Stage" = "#D34447")

gender_colors <- c("Male" = "#AACBFE", "Female" = "#DE8686")

my_color=c("#4675FE","#F9F5C9","#FCA605")

ann_ht1 <- HeatmapAnnotation(
                            Status=meta_scstage_df$stage,
                            Gender=meta_scstage_df$gender,
                            annotation_name_side = "left",
                            col=list(
                                  Status = status_colors,
                                  Gender = gender_colors))
sorted_h_dat$elemnets <- as.factor(sorted_h_dat$elemnets)

row_enhancer <- c("None"="#F2F2F2","CTCF_binding_site"="#8DD3C7","enhancer"="#BEBADA",
                  "open_chromatin_region"="#FB8072","promoter"="#80B1D3","TF_binding_site"="#CCEBC5")


row_ann_element <- rowAnnotation(Elements=sorted_h_dat$elemnets,col=list(Elements=row_enhancer))

row_groups <- c(rep("GO:0098742;GO:0034329;GO:0007043;GO:0030334", 3),
                rep("GO:0098742;GO:0034329;GO:0007043",18),
                rep("GO:0034329;GO:0007043;GO:0030334",3),
                rep("GO:0098742;GO:0034329",9),
                rep("GO:0098742;GO:0007043",5),
                rep("GO:0007043;GO:0030334",1),
                rep("GO:0098742;GO:0030334",2),
                rep("GO:0098742",26),
                rep("GO:0034329",1),
                rep("GO:0007043",3),
                rep("GO:0030334",53))

row_annot <- rowAnnotation(GO = row_groups,
                           col = list(
                           GO=c("GO:0034329"="#F9F5C9",
                                "GO:0030334"= "#63A6D0",
                                "GO:0007043"="#4675FE",
                                "GO:0098742"="#DE8686",
                                "GO:0007043;GO:0030334"="#6C040F",
                                "GO:0034329;GO:0007043;GO:0030334"="#FCBC85",
                                "GO:0098742;GO:0034329"="#F96949",
                                "GO:0098742;GO:0007043" ="#FCEBDE",
                                "GO:0098742;GO:0030334"="#AACBFE",
                                "GO:0098742;GO:0034329;GO:0007043" = "#FC8B39",
                                "GO:0098742;GO:0034329;GO:0007043;GO:0030334" = "#D94500")),                         
                                annotation_legend_param = list(
                                  GO = list(
                                    at=c(
                                    "GO:0034329",
                                    "GO:0030334",
                                    "GO:0007043",
                                    "GO:0007043;GO:0030334",
                                    "GO:0098742",
                                    "GO:0034329;GO:0007043;GO:0030334",
                                    "GO:0098742;GO:0034329",
                                    "GO:0098742;GO:0007043",
                                    "GO:0098742;GO:0030334",
                                    "GO:0098742;GO:0034329;GO:0007043",
                                    "GO:0098742;GO:0034329;GO:0007043;GO:0030334"
                                  ),
labels=c("Cell Junction Assembly",
         "Regulation Of Cell Migration",
         "Cell-Cell Junction Assembly",
         "Cell-Cell Junction Assembly;Regulation Of Cell Migration",
         "Cell-Cell Adhesion Via Plasma-Membrane Adhesion Molecules",
         "Cell Junction Assembly;Cell-Cell Junction Assembly;Regulation Of Cell Migration",
         "Cell-Cell Adhesion Via Plasma-Membrane Adhesion Molecules;Cell Junction Assembly",
         "Cell-Cell Adhesion Via Plasma-Membrane Adhesion Molecules;Cell-Cell Junction Assembly",
         "Cell-Cell Adhesion Via Plasma-Membrane Adhesion Molecules;Regulation Of Cell Migration",
         "Cell-Cell Adhesion Via Plasma-Membrane Adhesion Molecules;Cell Junction Assembly;Cell-Cell Junction Assembly",
"Cell Junction Assembly;Cell-Cell Junction Assembly;Regulation Of Cell Migration;Cell-Cell Adhesion Via Plasma-Membrane Adhesion Molecules")
)
))

heat_map_1 <- Heatmap(heatmap_dat,
                      cluster_rows = T,
                      cluster_columns = T,
                      row_split = row_groups,
                      right_annotation = c(row_annot,row_ann_element),
                      top_annotation = ann_ht1,
                      show_column_names = F,
                     show_row_dend = F,
                     show_row_names = T,
                     row_title = NULL,
                    col = colorRampPalette(my_color)(100),
                    na_col = "white",
                    width =unit(20,"cm"),
                     height = unit(38,"cm")
                      )
heat_map_1
# pdf(paste0(fig_dir,"figS5E_scstage_topGO_go_genes_20240802.pdf"),width = 15,height = 25)
draw(heat_map_1, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")

# dev.off()


```

## technical
```{r}
sessionInfo()
```

